### PR TITLE
Pipes -- Diane Lenertz -- hotel

### PR DESCRIPTION
**Pull request by lenertzdiane**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   I had to decide to keep the block class even though all it did was initialize and make a block object. I couldn't see a way of doing this that would be cleaner and I figured that it was better to have more separation than less on this project   |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  I definitely gained more clarity on oo design and the single responsibility principle/refactoring during the course of this assignment. |
| Describe a _nominal test_ that you wrote for this assignment. |  when someone makes a reservation for two days  |
| Describe an _edge case test_ that you wrote for this assignment. |  when someone puts in a beginning date after an end date |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I did really well with this for the first two waves but the last wave I got out of order (and it created problems) because I was excited and curious about making modifications for blocks |

### CODE REVIEW

File: lib/hotel.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+      def make_block_reservation(partyname)
+        id = reservations.length + 1
+        block = find_block(partyname)
+        date1 = block.daterange.beginning
+        date2 = block.daterange.ending
+        cost = block.discountedrate
+        block.rooms.each do |room|
+          if is_available?(room, date1, date2, partyname)
+            reservation =  Reservation.new(id, room, cost, date1, date2, partyname)
```
Comment (droberts-sea) [inline lib/hotel.rb:41]: This is an example of tight coupling. Instead of directly inspecting the block's list of rooms, it might be cleaner to write a method `Block#select_room` that does the work. That way, if the implementation of `Block` changes, this code will not have to change.

From your comprehension questions, it seems like you had a sense that this was the case - this sort of behavior is what would justify having a separate `Block` class.

File: lib/daterange.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      elsif (one_day? && date == @beginning) || (@beginning <= date && date <= @ending)
+        return true
+      else
+        return false
+      end
+    end
+
+
+    def overlap?(daterange)
+      return true if include?(daterange.beginning) || include?(daterange.ending)
```
Comment (droberts-sea) [inline lib/daterange.rb:50]: I really like this design! Implementing `overlap?` and `include?` here in `DateRange` makes that functionality much easier to test in isolation, and helps keep your `Hotel` code very readable.

File: lib/hotel.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+      def make_block_rooms(number_rooms, date1, date2)
+        rooms_in_block = []
+        index = 0
+        until rooms_in_block.length == number_rooms
+          if is_available?(@rooms[index], date1, date2)
+            rooms_in_block << @rooms[index]
+          end
+          index += 1
+          raise BookingError.new('There are no rooms available to make a block') if index > 20
```
Comment (droberts-sea) [inline lib/hotel.rb:137]: I believe this will result in an error if room 20 is the last room assigned to the block.

File: lib/hotel.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        end
+        is_in_block?(room_number, date1, date2, partyname, thisdaterange)
+      end
+
+      #IS THIS BETTER SEPARATED?
+      def is_in_block?(room_number, date1, date2=nil, partyname = nil, thisdaterange)
+        if partyname == nil
+          @blocks.each do |block|
+            if block.rooms.include? room_number
+              if block.daterange.overlap? thisdaterange
```
Comment (droberts-sea) [inline lib/hotel.rb:113]: Two observations. First, it looks like you're not using all the arguments here. `partyname` must be `nil` for the method to do anything, and `date1` and `date2` are ignored completely.

Second, this is another place where you've got tight coupling between `Hotel` and `Block`. If you had a method `Block#matches?(room_number, date_range)`, then this method could be shortened to:

```ruby
def is_in_block?(room_number, thisdaterange)
  return ! @blocks.any? { |block| block.matches?(room_number, thisdaterange) }
end
```

and it would be more resilient to changes to `Block`.

File: specs/daterange_spec.rb [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      @daterange3 = Hotel::DateRange.new(@date3, @date5)
+     @daterange6 = Hotel::DateRange.new(@date6, @date7)
+    end
+    it 'returns true if the date range overlaps with the passed date range by one or all dates' do
+      @daterange.overlap?(@daterange2).must_equal true
+      @daterange.overlap?(@daterange3).must_equal true
+    end
+    it 'returns false if the date range does not overlap with the passed date range' do
+      @daterange.overlap?(@daterange6).must_equal false
+    end
```
Comment (droberts-sea) [inline specs/daterange_spec.rb:148]: This is a good start, but there are many interesting cases here you're not covering. My complete list includes:
- Does Overlap:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Does Not Overlap:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date

File: specs/hotel_spec.rb [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      @hotel.make_reservation(date1: @date1, date2: @date2).must_be_kind_of Array
+    end
+    it 'creates a reservation object with one date and pushes to reservations' do
+      @hotel.make_reservation(date1: @date1).must_be_kind_of Array
+    end
+    it 'returns an array of reservation objects' do
+      @hotel.reservations.each do |reservation|
+        reservation.must_be_instance_of Hotel::Reservation
+      end
+    end
```
Comment (droberts-sea) [inline specs/hotel_spec.rb:38]: You should test your error handling here. What happens if there aren't any rooms available?

File: specs/hotel_spec.rb [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      @hotel.view_available(date1: @date1).must_equal [:room2, :room3, :room4, :room5, :room6, :room7, :room8, :room9, :room10,
+        :room11, :room12, :room13, :room14, :room15, :room16, :room17, :room18, :room19, :room20]
+      @hotel.view_available(date1: @date1, date2: @date3).must_equal [:room2, :room3, :room4, :room5, :room6, :room7, :room8, :room9, :room10,
+        :room11, :room12, :room13, :room14, :room15, :room16, :room17, :room18, :room19, :room20]
+      @hotel.view_available(date1: @date3, date2: @date4).must_equal [:room2, :room3, :room4, :room5, :room6, :room7, :room8, :room9, :room10,
+        :room11, :room12, :room13, :room14, :room15, :room16, :room17, :room18, :room19, :room20]
+      @hotel.make_reservation(date1: @date1, date2: @date2)
+      @hotel.view_available(date1: @date1, date2: @date2).must_equal [:room3, :room4, :room5, :room6, :room7, :room8, :room9, :room10,
+        :room11, :room12, :room13, :room14, :room15, :room16, :room17, :room18, :room19, :room20]
+    end
```
Comment (droberts-sea) [inline specs/hotel_spec.rb:54]: Other cases I'd be interested in:
- What if there are no reservations? (it should return a list of all the rooms)
- Do reservations that don't overlap the date range affect the list? (they shouldn't)
- What if there are no vacancies? (it should return an empty array)

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes - good work!
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | mostly - see inline comment
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | good start - see inline comments
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | good start - see inline
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | yes
**Additional Feedback** |

Great job overall! Though there are a few places where functionality could be moved out of `Hotel` and into other classes, your high level design seems deliberate and well-thought-out. The actual Ruby code looks quite good, and though there are some places where test coverage could be expanded, this is a good start. Keep up the hard work!
